File: devices/backend/src/app/core/reliability/circuit_breaker.py
"""Circuit breaker pattern implementation for fault tolerance."""

import asyncio
import logging
import time
from collections.abc import Callable
from dataclasses import dataclass
from enum import Enum
from typing import Any

logger = logging.getLogger(__name__)


class CircuitState(Enum):
    """Circuit breaker states."""

    CLOSED = "closed"  # Normal operation, requests pass through
    OPEN = "open"  # Circuit is open, requests fail immediately
    HALF_OPEN = "half_open"  # Testing if service recovered


class CircuitBreakerError(Exception):
    """Raised when circuit breaker is open."""



@dataclass
class CircuitBreakerConfig:
    """Configuration for circuit breaker behavior."""

    max_failures: int = 5
    timeout: float = 30.0  # seconds before transitioning from Open to Half-Open
    max_half_open_requests: int = 3


class CircuitBreaker:
    """
    Circuit breaker implementation for protecting against cascading failures.

    Usage:
        breaker = CircuitBreaker("my-service")
        result = await breaker.call(my_async_function)
    """

    def __init__(
        self,
        name: str,
        config: CircuitBreakerConfig | None = None,
    ):
        """
        Initialize circuit breaker.

        Args:
            name: Name for logging and identification
            config: Configuration (uses defaults if not provided)
        """
        self.name = name
        self.config = config or CircuitBreakerConfig()
        self.state = CircuitState.CLOSED
        self.failures = 0
        self.last_failure_time = 0.0
        self.half_open_requests = 0
        self.half_open_successes = 0
        self._lock = asyncio.Lock()

    async def call(self, operation: Callable[[], Any]) -> Any:
        """
        Execute an async operation through the circuit breaker.

        Args:
            operation: Async function to execute

        Returns:
            Result from the operation

        Raises:
            CircuitBreakerError: If circuit is open
            Exception: Any exception from the operation
        """
        if not await self._allow_request():
            raise CircuitBreakerError(f"Circuit breaker '{self.name}' is open")

        try:
            result = await operation()
            await self._on_success()
            return result
        except Exception:
            await self._on_failure()
            raise

    async def _allow_request(self) -> bool:
        """Check if a request should be allowed based on current state."""
        async with self._lock:
            now = time.time()

            if self.state == CircuitState.CLOSED:
                return True

            if self.state == CircuitState.OPEN:
                # Check if we should transition to Half-Open
                if now - self.last_failure_time >= self.config.timeout:
                    logger.info(
                        f"Circuit breaker '{self.name}': Transitioning from Open to Half-Open"
                    )
                    self.state = CircuitState.HALF_OPEN
                    self.half_open_requests = 0
                    self.half_open_successes = 0
                    return True
                return False

            if self.state == CircuitState.HALF_OPEN:
                # Allow limited requests in Half-Open state
                if self.half_open_requests < self.config.max_half_open_requests:
                    self.half_open_requests += 1
                    return True
                return False

            return False

    async def _on_success(self) -> None:
        """Handle a successful operation."""
        async with self._lock:
            if self.state == CircuitState.CLOSED:
                # Reset failure count on success in Closed state
                self.failures = 0

            elif self.state == CircuitState.HALF_OPEN:
                self.half_open_successes += 1
                # If we've had enough successes, close the circuit
                if self.half_open_successes >= self.config.max_half_open_requests:
                    logger.info(
                        f"Circuit breaker '{self.name}': Transitioning from Half-Open to Closed"
                    )
                    self.state = CircuitState.CLOSED
                    self.failures = 0
                    self.half_open_requests = 0
                    self.half_open_successes = 0

    async def _on_failure(self) -> None:
        """Handle a failed operation."""
        async with self._lock:
            self.failures += 1
            self.last_failure_time = time.time()

            if self.state == CircuitState.CLOSED:
                if self.failures >= self.config.max_failures:
                    logger.warning(
                        f"Circuit breaker '{self.name}': Transitioning from Closed to Open "
                        f"after {self.failures} failures"
                    )
                    self.state = CircuitState.OPEN

            elif self.state == CircuitState.HALF_OPEN:
                # Any failure in Half-Open state reopens the circuit
                logger.warning(
                    f"Circuit breaker '{self.name}': Transitioning from Half-Open to Open "
                    f"due to failure"
                )
                self.state = CircuitState.OPEN
                self.half_open_requests = 0
                self.half_open_successes = 0

    def get_state(self) -> CircuitState:
        """Get current circuit breaker state."""
        return self.state

    async def reset(self) -> None:
        """Manually reset circuit breaker to Closed state."""
        async with self._lock:
            logger.info(f"Circuit breaker '{self.name}': Manual reset to Closed state")
            self.state = CircuitState.CLOSED
            self.failures = 0
            self.half_open_requests = 0
            self.half_open_successes = 0

File: mentor/backend/src/reliability/circuit_breaker.go
package reliability

import (
	"errors"
	"fmt"
	"log"
	"sync"
	"time"
)

var (
	// ErrCircuitOpen is returned when the circuit breaker is open
	ErrCircuitOpen = errors.New("circuit breaker is open")
)

// CircuitState represents the state of a circuit breaker
type CircuitState int

const (
	// StateClosed means requests pass through normally
	StateClosed CircuitState = iota
	// StateOpen means requests are rejected immediately
	StateOpen
	// StateHalfOpen means limited requests are allowed to test if the service recovered
	StateHalfOpen
)

func (s CircuitState) String() string {
	switch s {
	case StateClosed:
		return "closed"
	case StateOpen:
		return "open"
	case StateHalfOpen:
		return "half-open"
	default:
		return "unknown"
	}
}

// CircuitBreakerConfig defines configuration for a circuit breaker
type CircuitBreakerConfig struct {
	// MaxFailures is the number of consecutive failures before opening the circuit
	MaxFailures int
	// Timeout is how long to wait before transitioning from Open to Half-Open
	Timeout time.Duration
	// MaxHalfOpenRequests is the number of requests to allow in Half-Open state
	MaxHalfOpenRequests int
}

// DefaultCircuitBreakerConfig returns sensible defaults
func DefaultCircuitBreakerConfig() CircuitBreakerConfig {
	return CircuitBreakerConfig{
		MaxFailures:         5,
		Timeout:             30 * time.Second,
		MaxHalfOpenRequests: 3,
	}
}

// CircuitBreaker implements the circuit breaker pattern
type CircuitBreaker struct {
	config            CircuitBreakerConfig
	state             CircuitState
	failures          int
	lastFailureTime   time.Time
	halfOpenRequests  int
	halfOpenSuccesses int
	mu                sync.RWMutex
	name              string
}

// NewCircuitBreaker creates a new circuit breaker with the given configuration
func NewCircuitBreaker(name string, config CircuitBreakerConfig) *CircuitBreaker {
	return &CircuitBreaker{
		config: config,
		state:  StateClosed,
		name:   name,
	}
}

// Execute runs the given operation through the circuit breaker
func (cb *CircuitBreaker) Execute(operation func() error) error {
	// Check if we should allow the request
	if !cb.allowRequest() {
		return fmt.Errorf("%w: circuit breaker '%s' is open", ErrCircuitOpen, cb.name)
	}

	// Execute the operation
	err := operation()

	// Record the result
	cb.recordResult(err)

	return err
}

// allowRequest checks if a request should be allowed based on the current state
func (cb *CircuitBreaker) allowRequest() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	now := time.Now()

	switch cb.state {
	case StateClosed:
		return true

	case StateOpen:
		// Check if we should transition to Half-Open
		if now.Sub(cb.lastFailureTime) >= cb.config.Timeout {
			log.Printf("Circuit breaker '%s': Transitioning from Open to Half-Open", cb.name)
			cb.state = StateHalfOpen
			cb.halfOpenRequests = 0
			cb.halfOpenSuccesses = 0
			return true
		}
		return false

	case StateHalfOpen:
		// Allow limited requests in Half-Open state
		if cb.halfOpenRequests < cb.config.MaxHalfOpenRequests {
			cb.halfOpenRequests++
			return true
		}
		return false

	default:
		return false
	}
}

// recordResult updates the circuit breaker state based on the operation result
func (cb *CircuitBreaker) recordResult(err error) {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if err != nil {
		cb.onFailure()
	} else {
		cb.onSuccess()
	}
}

// onSuccess handles a successful operation
func (cb *CircuitBreaker) onSuccess() {
	switch cb.state {
	case StateClosed:
		// Reset failure count on success in Closed state
		cb.failures = 0

	case StateHalfOpen:
		cb.halfOpenSuccesses++
		// If we've had enough successes, close the circuit
		if cb.halfOpenSuccesses >= cb.config.MaxHalfOpenRequests {
			log.Printf("Circuit breaker '%s': Transitioning from Half-Open to Closed", cb.name)
			cb.state = StateClosed
			cb.failures = 0
			cb.halfOpenRequests = 0
			cb.halfOpenSuccesses = 0
		}
	}
}

// onFailure handles a failed operation
func (cb *CircuitBreaker) onFailure() {
	cb.failures++
	cb.lastFailureTime = time.Now()

	switch cb.state {
	case StateClosed:
		if cb.failures >= cb.config.MaxFailures {
			log.Printf("Circuit breaker '%s': Transitioning from Closed to Open after %d failures",
				cb.name, cb.failures)
			cb.state = StateOpen
		}

	case StateHalfOpen:
		// Any failure in Half-Open state reopens the circuit
		log.Printf("Circuit breaker '%s': Transitioning from Half-Open to Open due to failure", cb.name)
		cb.state = StateOpen
		cb.halfOpenRequests = 0
		cb.halfOpenSuccesses = 0
	}
}

// State returns the current state of the circuit breaker
func (cb *CircuitBreaker) State() CircuitState {
	cb.mu.RLock()
	defer cb.mu.RUnlock()
	return cb.state
}

// Reset resets the circuit breaker to Closed state
func (cb *CircuitBreaker) Reset() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	log.Printf("Circuit breaker '%s': Manual reset to Closed state", cb.name)
	cb.state = StateClosed
	cb.failures = 0
	cb.halfOpenRequests = 0
	cb.halfOpenSuccesses = 0
}
